projects/com.oracle.truffle.llvm.types/src/com/oracle/truffle/llvm/types/LLVMFunction.java

```diff
@@ -216,4 +289,57 @@ public boolean equals(Object obj) {
         }
     }
 
+    private static class ForeignCallNode extends RootNode {
+
+        private final LLVMStack stack;
+
+        @Child private IndirectCallNode callNode;
+
+        protected ForeignCallNode() {
+            super(getLLVMLanguage(), null, new FrameDescriptor());
+            stack = getLLVMStack();
+            callNode = Truffle.getRuntime().createIndirectCallNode();
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            assert ForeignAccess.getArguments(frame).isEmpty();
+            final LLVMFunction function = (LLVMFunction) ForeignAccess.getReceiver(frame);
+            final CallTarget callTarget = getCallTarget(function);
+            return callNode.call(frame, callTarget, new Object[]{stack.allocate()});
```

Comment:
I believe that the stack is not freed any more and will result in a memory leak. Can we implement a getter for `LLVMStack#upperBounds` (and `#lowerBounds` for consistency) so that we use the already allocated stack memory? Currently the stack is freed in `LLVMGlobalRootNode`. Is this node used anywhere? If not, then global variables are probably not yet supported.